### PR TITLE
Add cash payment method (SHUUP-2723)

### PR DESCRIPTION
**Pull request by gurch101**
Add cash payments to the custom payment processor which sets created orders
to fully paid on creation. Also add rounding behavior component to round
order totals to the nearest 0.05.

Refs SHUUP-2723

### CODE REVIEW

File: shoop/core/models/_service_behavior.py [hunk 1]
```diff
@@ -164,3 +167,42 @@ def get_unavailability_reasons(self, service, source):
...
+
+class RoundingBehaviorComponent(ServiceBehaviorComponent):
+    name = _("Rounding")
+    help_text = _("Round total order price to the nearest five cents.")
+
+    rounding_behavior = EnumIntegerField(
+        RoundingBehavior,
+        default=RoundingBehavior.ROUND_DOWN,
+        verbose_name=_("rounding behavior"))
+
```
Comment (shawnadelic) [inline shoop/core/models/_service_behavior.py]: This behavior component should probably not go in core, but should be provided by the AppConfig.

File: shoop/admin/modules/orders/receivers.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            payment_identifier="Cash-%s" % now().isoformat(),
+            description="Cash Payment"
+        )
+
+
+def handle_custom_payment_return_requests(sender, order, *args, **kwargs):
+    payment_processor = order.payment_method.payment_processor if order.payment_method else None
+    if isinstance(payment_processor, CustomPaymentProcessor):
+        service = order.payment_method.choice_identifier
+        if service == 'cash':
```
Comment (tulimaki) [inline shoop/admin/modules/orders/receivers.py]: Add handling also for "manual" payment service. Use double quotes for cash.

File: shoop/core/models/_service_behavior.py [hunk 3]
```diff
@@ -164,3 +167,42 @@ def get_unavailability_reasons(self, service, source):
...
+        rounded_amt = 0
+        if self.rounding_behavior == RoundingBehavior.ROUND_UP:
+            rounded_amt = self.round_up(total_price) - total_price
+        else:
+            rounded_amt = self.round_down(total_price) - total_price
+        rounded_amt = source.create_price(decimal.Decimal(rounded_amt))
+        yield ServiceCost(rounded_amt, self.rounding_behavior.label)
+
+    def get_unavailability_reasons(self, service, source):
+        if not source.creator or not source.creator.is_superuser:
```
Comment (tulimaki) [inline shoop/core/models/_service_behavior.py]: Allow also is_staff here.

File: shoop/core/models/_service_behavior.py [hunk 4]
```diff
@@ -168,3 +171,42 @@ def get_unavailability_reasons(self, service, source):
...
+    rounding_behavior = EnumIntegerField(
+        RoundingBehavior,
+        default=RoundingBehavior.ROUND_DOWN,
+        verbose_name=_("rounding behavior"))
+
+    def round_down(self, amount):
+        return (amount * 2).quantize(decimal.Decimal('.1'), rounding=decimal.ROUND_DOWN) / 2
+
+    def round_up(self, amount):
+        return (amount * 2).quantize(decimal.Decimal('.1'), rounding=decimal.ROUND_UP) / 2
```
Comment (suutari-ai) [inline shoop/core/models/_service_behavior.py]: The rounding "quant", which seems to be 0.05 here, should be configurable: Add it as field to the model. Also the so called "nickel rounding" algorithm could be useful as a separate util. You should probably define the nickel rounding algorithm in shoop.utils.numbers. It could be something like this:
```python
def nickel_round(value, quant=Decimal('0.05'), rounding=decimal.ROUND_HALF_UP):
    """
    Round decimal value to nearest quant.

    For example:

    >>> assert nickel_round(Decimal('10.31')) == Decimal('10.30')
    >>> assert nickel_round(Decimal('10.33')) == Decimal('10.35')
    >>> assert nickel_round(Decimal('10.325')) == Decimal('10.35')
    >>> assert nickel_round(Decimal('10.3249')) == Decimal('10.30')
    >>> assert nickel_round(Decimal('10.31'), Decimal('0.125')) == Decimal('10.250')
    >>> assert nickel_round(Decimal('10.32'), Decimal('0.125')) == Decimal('10.375')
    """
    assert isinstance(value, Decimal)
    return (value / quant).quantize(1, rounding=rounding) * quant
```

Comment (suutari-ai) [inline shoop/core/models/_service_behavior.py]: Btw. since our Money and Price classes are also subclasses of Decimal, the algorithm works fine with them too:
```python
In [12]: nickel_round(TaxfulPrice('10.241', 'EUR'))
Out[12]: TaxfulPrice('10.25', 'EUR')
```

Comment (suutari-ai) [inline shoop/core/models/_service_behavior.py]: I created PR #555 for adding that function. You may merge that and rebase your PR on top.

File: shoop/core/models/_service_behavior.py [hunk 5]
```diff
@@ -168,3 +172,39 @@ def get_unavailability_reasons(self, service, source):
...
+    class Labels:
+        ROUND_UP = _("round up")
+        ROUND_DOWN = _("round down")
+
+
+class RoundingBehaviorComponent(ServiceBehaviorComponent):
+    name = _("Rounding")
+    help_text = _("Round total order price to the nearest quant.")
+
+    rounding_quant = models.DecimalField(
```
Comment (tulimaki) [inline shoop/core/models/_service_behavior.py]: This is never used.

File: shoop/core/models/_service_behavior.py [hunk 6]
```diff
@@ -168,3 +172,39 @@ def get_unavailability_reasons(self, service, source):
...
+        max_digits=36, decimal_places=9,
+        verbose_name=_("rounding quant"))
```
Comment (suutari-ai) [inline shoop/core/models/_service_behavior.py]: I think `Decimal('0.05')` would be a good default value for the `rounding_quant`.

File: shoop/core/models/_service_behavior.py [hunk 7]
```diff
@@ -168,3 +172,39 @@ def get_unavailability_reasons(self, service, source):
...
+
+    def get_costs(self, service, source):
+        total_price = source.total_price_of_products.value
+        rounded_amt = 0
+        if self.rounding_behavior == RoundingBehavior.ROUND_UP:
+            rounded_amt = nickel_round(total_price, quant=self.rounding_quant, rounding=decimal.ROUND_UP)
+        else:
+            rounded_amt = nickel_round(total_price, quant=self.rounding_quant, rounding=decimal.ROUND_DOWN)
+        rounded_amt = source.create_price(decimal.Decimal(rounded_amt - total_price))
+        yield ServiceCost(rounded_amt, self.rounding_behavior.label)
```
Comment (suutari-ai) [inline shoop/core/models/_service_behavior.py]: 1. You're assigning three different values to the `rounded_amt` variable. That makes it little hard to follow. Use different variables for different things.
2. You can operate with prices, no need to drop to bare decimals. `nickel_round(x, quant)` returns a `Price` if `x` is a `Price` and `quant` can still be bare `Decimal`.
3. Are these rounding modes "correct"? At least in Finnish stores prices are rounded with ROUND_HALF_UP method, e.g. 2,12 € will be rounded to 2,10 € and 2,13 € to 2,15 €. Neither ROUND_UP or ROUND_DOWN does that. We should provide  at least ROUND_HALF_UP too, but wouldn't hurt to provide some others too as long as they're supported by the decimal module.


I think this could be as simple as this:
```python
def get_costs(self, service source):
    total_price = source.total_price_of_products
    rounded = nickel_round(total_price, self.quant, self.mode)
    remainder = rounded - total_price
    yield ServiceCost(remainder, self.mode.label)
```
I'm assuming that the enum values for rounding mode use directly values of `decimal.ROUND_UP` etc. to avoid mapping them.

File: shoop/core/models/_service_behavior.py [hunk 8]
```diff
@@ -168,3 +172,39 @@ def get_unavailability_reasons(self, service, source):
...
             yield ValidationError(_("Service is not available for any of the customers groups."))
+
+
+class RoundingBehavior(Enum):
+    ROUND_UP = 1
+    ROUND_DOWN = 2
+
+    class Labels:
+        ROUND_UP = _("round up")
+        ROUND_DOWN = _("round down")
```
Comment (suutari-ai) [inline shoop/core/models/_service_behavior.py]: Provide also ROUND_HALF_UP and ROUND_HALF_DOWN and use the values of `decimal.ROUND_*` constants directly to avoid mapping those. They are strings , so the EnumIntegerField below has to be changed to EnumField too. Python's decimal module calls this rounding mode and IMHO that's better name than rounding behavior.

```python
class RoundingMode(Enum):
    ROUND_HALF_UP = decimal.ROUND_HALF_UP
    ROUND_HALF_DOWN = decimal.ROUND_HALF_DOWN
    ROUND_UP = decimal.ROUND_UP
    ...
    class Labels:
        ROUND_HALF_UP = _("round to nearest with ties going away from zero")
        ...
```

I don't think using those labels directly for the description of the cost is a good idea though...

**Comment by suutari-ai** [comment]
Good job, @gurch101! Thanks!
